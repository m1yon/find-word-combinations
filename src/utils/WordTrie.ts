export class WordTrie {
  private value?: string;
  private children: WordTrie[];
  private isWord: boolean;

  constructor(args?: { value?: string; isWord?: boolean }) {
    this.value = args?.value;
    this.children = [];
    this.isWord = args?.isWord ?? false;
  }

  addWord(word: string) {
    let node = this as WordTrie;

    word.split("").forEach((character, index) => {
      const nextChildNode = node.getChildren().find((child) => {
        return child.getValue() === character;
      });

      if (nextChildNode) {
        node = nextChildNode;
        return;
      }

      const isLastCharacter = index === word.length - 1;

      const newNode = new WordTrie({
        value: character,
        isWord: isLastCharacter,
      });

      node = node.addChild(newNode);
    });
  }

  addWords(words: string[]) {
    for (const word of words) {
      this.addWord(word);
    }
  }

  getValue() {
    return this.value;
  }

  getChildren() {
    return this.children;
  }

  getIsWord() {
    return this.isWord;
  }

  setIsWord(isWord: boolean) {
    this.isWord = isWord;
  }

  addChild(node: WordTrie) {
    this.children.push(node);
    return node;
  }
}
