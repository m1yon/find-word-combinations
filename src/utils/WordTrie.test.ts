import { expect, it } from "bun:test";
import { WordTrie } from "./WordTrie";

it("generates a single node correctly", () => {
  const root = new WordTrie({ value: "a" });

  expect(root).toMatchSnapshot();
});

it("generates the correct tree given a single word", () => {
  const root = new WordTrie({ value: "" });
  root.addWord("great");

  expect(root).toMatchSnapshot();
});

it("generates the correct tree given multiple words that do not connect", () => {
  const root = new WordTrie({ value: "" });
  root.addWord("red");
  root.addWord("tree");
  root.addWord("far");

  expect(root).toMatchSnapshot();
});
