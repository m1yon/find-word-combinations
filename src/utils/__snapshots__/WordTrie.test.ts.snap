// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generates the correct tree given a single word 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [
                    WordTrie {
                      "children": [],
                      "isWord": true,
                      "value": "t",
                    },
                  ],
                  "isWord": false,
                  "value": "a",
                },
              ],
              "isWord": false,
              "value": "e",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "g",
    },
  ],
  "isWord": false,
  "value": "",
}
`;

exports[`generates a single node correctly 1`] = `
WordTrie {
  "children": [],
  "isWord": false,
  "value": "a",
}
`;

exports[`generates the correct tree given multiple words 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "d",
            },
          ],
          "isWord": false,
          "value": "e",
        },
      ],
      "isWord": false,
      "value": "r",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [],
                  "isWord": true,
                  "value": "e",
                },
              ],
              "isWord": false,
              "value": "e",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "t",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "r",
            },
          ],
          "isWord": false,
          "value": "a",
        },
      ],
      "isWord": false,
      "value": "f",
    },
  ],
  "isWord": false,
  "value": "",
}
`;

exports[`generates the correct tree given multiple words that do not connect 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "d",
            },
          ],
          "isWord": false,
          "value": "e",
        },
      ],
      "isWord": false,
      "value": "r",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [],
                  "isWord": true,
                  "value": "e",
                },
              ],
              "isWord": false,
              "value": "e",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "t",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "r",
            },
          ],
          "isWord": false,
          "value": "a",
        },
      ],
      "isWord": false,
      "value": "f",
    },
  ],
  "isWord": false,
  "value": "",
}
`;

exports[`generates the correct tree given multiple words that do NOT connect 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "d",
            },
          ],
          "isWord": false,
          "value": "e",
        },
      ],
      "isWord": false,
      "value": "r",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [],
                  "isWord": true,
                  "value": "e",
                },
              ],
              "isWord": false,
              "value": "e",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "t",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "r",
            },
          ],
          "isWord": false,
          "value": "a",
        },
      ],
      "isWord": false,
      "value": "f",
    },
  ],
  "isWord": false,
  "value": "",
}
`;

exports[`generates the correct tree given multiple words that do connect 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [
                    WordTrie {
                      "children": [],
                      "isWord": true,
                      "value": "t",
                    },
                  ],
                  "isWord": false,
                  "value": "a",
                },
              ],
              "isWord": false,
              "value": "e",
            },
            WordTrie {
              "children": [
                WordTrie {
                  "children": [
                    WordTrie {
                      "children": [
                        WordTrie {
                          "children": [
                            WordTrie {
                              "children": [],
                              "isWord": true,
                              "value": "r",
                            },
                          ],
                          "isWord": false,
                          "value": "e",
                        },
                      ],
                      "isWord": true,
                      "value": "l",
                    },
                  ],
                  "isWord": false,
                  "value": "w",
                },
              ],
              "isWord": false,
              "value": "o",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "g",
    },
  ],
  "isWord": false,
  "value": "",
}
`;

exports[`generates the correct trie given a single word 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [
                    WordTrie {
                      "children": [],
                      "isWord": true,
                      "value": "t",
                    },
                  ],
                  "isWord": false,
                  "value": "a",
                },
              ],
              "isWord": false,
              "value": "e",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "g",
    },
  ],
  "isWord": false,
  "value": "",
}
`;

exports[`generates the correct trie given multiple words that do not connect 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "d",
            },
          ],
          "isWord": false,
          "value": "e",
        },
      ],
      "isWord": false,
      "value": "r",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [],
                  "isWord": true,
                  "value": "e",
                },
              ],
              "isWord": false,
              "value": "e",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "t",
    },
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [],
              "isWord": true,
              "value": "r",
            },
          ],
          "isWord": false,
          "value": "a",
        },
      ],
      "isWord": false,
      "value": "f",
    },
  ],
  "isWord": false,
  "value": "",
}
`;

exports[`generates the correct trie given multiple words that do connect 1`] = `
WordTrie {
  "children": [
    WordTrie {
      "children": [
        WordTrie {
          "children": [
            WordTrie {
              "children": [
                WordTrie {
                  "children": [
                    WordTrie {
                      "children": [],
                      "isWord": true,
                      "value": "t",
                    },
                  ],
                  "isWord": false,
                  "value": "a",
                },
              ],
              "isWord": false,
              "value": "e",
            },
            WordTrie {
              "children": [
                WordTrie {
                  "children": [
                    WordTrie {
                      "children": [
                        WordTrie {
                          "children": [
                            WordTrie {
                              "children": [],
                              "isWord": true,
                              "value": "r",
                            },
                          ],
                          "isWord": false,
                          "value": "e",
                        },
                      ],
                      "isWord": true,
                      "value": "l",
                    },
                  ],
                  "isWord": false,
                  "value": "w",
                },
              ],
              "isWord": false,
              "value": "o",
            },
          ],
          "isWord": false,
          "value": "r",
        },
      ],
      "isWord": false,
      "value": "g",
    },
  ],
  "isWord": false,
  "value": "",
}
`;
